src/v/kafka/server/tests/quota_manager_bench.cc

```diff
@@ -78,59 +81,384 @@ ss::future<> test_quota_manager(size_t count, bool use_unique) {
+    return run_tc(throughput_test_case{
       .fetch_tp = 1000,
       .use_unique = false,
     });
 }
 
-PERF_TEST_C(quota_group, test_quota_manager_on_limited_unique) {
-    co_await run_tc(test_case{
+PERF_TEST_CN(throughput_group, test_quota_manager_on_limited_unique) {
+    return run_tc(throughput_test_case{
       .fetch_tp = 1000,
       .use_unique = true,
     });
 }
 
-PERF_TEST_C(quota_group, test_quota_manager_off_shared) {
-    co_await run_tc(test_case{
+PERF_TEST_CN(throughput_group, test_quota_manager_off_shared) {
+    return run_tc(throughput_test_case{
       .fetch_tp = std::nullopt,
       .use_unique = false,
     });
 }
 
-PERF_TEST_C(quota_group, test_quota_manager_off_unique) {
-    co_await run_tc(test_case{
+PERF_TEST_CN(throughput_group, test_quota_manager_off_unique) {
+    return run_tc(throughput_test_case{
       .fetch_tp = std::nullopt,
       .use_unique = true,
     });
 }
+
+struct latency_test_case {
+    bool is_new_client;
+    bool is_produce_not_fetch;
+    int n_other_clients;
+    bool on_shard_0;
+    bool no_quotas;
+};
+
+const static size_t n_repeats = 100;
+
+future<size_t> run_latency_test(latency_test_case tc) {
+    ss::sharded<cluster::client_quota::store> quota_store;
+    kafka::quota_manager::client_quotas_t buckets_map;
+    ss::sharded<kafka::quota_manager> sqm;
+    co_await quota_store.start();
+    co_await sqm.start(std::ref(buckets_map), std::ref(quota_store));
+    co_await sqm.invoke_on_all(&kafka::quota_manager::start);
+
+    auto shard = tc.on_shard_0 ? 0 : 1;
+
+    co_await ss::smp::submit_to(
+      shard,
+      seastar::coroutine::lambda([&sqm, &quota_store, tc]() -> ss::future<> {
+          auto& qm = sqm.local();
+
+          // Try to create a realistic setup
+          if (!tc.no_quotas) {
+              using cluster::client_quota::entity_key;
+              using cluster::client_quota::entity_value;
+              auto key = entity_key{entity_key::client_id_default_match{}};
+              auto value = entity_value{
+                .producer_byte_rate = 1 << 30, .consumer_byte_rate = 1 << 30};
+              co_await quota_store.invoke_on_all(
+                [&key, &value](cluster::client_quota::store& qs) {
+                    qs.set_quota(key, value);
+                });
+          }
+
+          auto now = kafka::quota_manager::clock::now();
+
+          // Have a non-trivial number of existing clients in the map
+          for (int i = 0; i < tc.n_other_clients; i++) {
+              co_await qm.record_produce_tp_and_throttle(
+                fmt::format("client-{}", i), 1, now);
+          }
+
+          // Pre-generate the client-id's used during the benchmark
+          auto client_ids = std::vector<ss::sstring>{};
+          client_ids.reserve(n_repeats);
+          for (int i = 0; i < n_repeats; i++) {
+              auto& id = client_ids.emplace_back(
+                tc.is_new_client ? fmt::format("new-client-{}", i)
```

Comment:
If `is_new_client` is false then we'll still emplace into the map in all the n iterations because we're creating a vector of `n_repeats` elements here. But the call to `record_produce_tp_and_throttle` a few lines below will have "no effect" in that the `fixed_client_id` will already be in the map.

Reply:
Oh this is a vector of course, me being dumb. Had a map in mind.